wagtail/models/__init__.py

```diff
@@ -3031,25 +3135,55 @@ def publishable_pages(self):
 
         publishable_pages = Page.objects.none()
 
-        for perm in self.permissions.filter(permission_type="publish"):
-            # user has publish permission on any subpage of perm.page
-            # (including perm.page itself)
-            publishable_pages |= Page.objects.descendant_of(perm.page, inclusive=True)
+        for page in self.get_pages_for_perms(["publish"]):
+            # user has publish permission on any subpage of page
+            # (including page itself)
+            publishable_pages |= Page.objects.descendant_of(page, inclusive=True)
 
         return publishable_pages
 
-    def can_publish_pages(self):
-        """Return True if the user has permission to publish any pages"""
+    @cached_property
+    def _can_publish_pages(self):
+        """Return True if the user has permission to publish any pages."""
         return self.publishable_pages().exists()
 
+    def can_publish_pages(self):
+        return self._can_publish_pages
+
     def can_remove_locks(self):
-        """Returns True if the user has permission to unlock pages they have not locked"""
+        """
+        Returns True if the user has permission
+        to unlock pages they have not locked.
+        """
         if self.user.is_superuser:
             return True
         if not self.user.is_active:
             return False
         else:
-            return self.permissions.filter(permission_type="unlock").exists()
+            return "unlock" in self.perm_types
+
+    @cached_property
+    def _site_details(self):
+        root_page = self.explorable_root_page()
+        if root_page:
+            root_site = root_page.get_site()
+        else:
+            root_site = None
+        real_site_name = None
+        if root_site:
+            real_site_name = (
+                root_site.site_name if root_site.site_name else root_site.hostname
+            )
+        return {
+            "root_page": root_page,
+            "root_site": root_site,
+            "site_name": real_site_name
+            if real_site_name
+            else settings.WAGTAIL_SITE_NAME,
+        }
```

Comment:
Nit: The ternary in the dict is difficult to read. We can avoid defining it as `None` too.
```suggestion
        real_site_name = settings.WAGTAIL_SITE_NAME
        if root_site:
            real_site_name = (
                root_site.site_name if root_site.site_name else root_site.hostname
            )
        return {
            "root_page": root_page,
            "root_site": root_site,
            "site_name": real_site_name,
        }
```